File: 05_arrays/05_19_pascals_traingle/PascalsTraingle.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class PascalsTraingle {
    public static void main(String[] args) {
        List<List<Integer>> result = generate(5);
        System.out.println(Arrays.toString(result.toArray()));
    }

    private static List<List<Integer>> generate(int numRows) {
        List<List<Integer>> result = new ArrayList<>();
        if (numRows == 0) {
            return result;
        }

        result.add(List.of(1));
        for (int i = 1; i < numRows; i++) {
            List<Integer> prevRow = result.get(i - 1);
            List<Integer> currRow = new ArrayList<>();
            currRow.add(1);

            for (int j = 1; j < i; j++) {
                currRow.add(prevRow.get(j - 1) + prevRow.get(j));
            }

            currRow.add(1);
            result.add(currRow);
        }

        return result;
    }
}

File: 05_arrays/05_19_pascals_traingle/pascals_traingle.py
def generate(numRows: int):
    result = []
    if numRows == 0:
        return result
    
    result.append([1])
    for i in range(1, numRows):
        prevRow = result[i - 1]
        currentRow = [1] * (i + 1)
        for j in range(1, i):
            currentRow[j] = prevRow[j - 1] + prevRow[j]
        result.append(currentRow)
    
    return result


print(generate(5))